### PR TITLE
support kernel annotations for versioned vmlinux and kernel modules in collapse-perf

### PR DESCRIPTION
**Pull request by AnderEnder**
Support kernel annotations for versioned vmlinux and kernel modules

Backport of https://github.com/brendangregg/FlameGraph/pull/232

Example of stack trace:
```
swapper     0 [012] 2281339.604067:   20834821 cycles:
        ffffffffc09ace93 bpf_prog_2b956549c660136a_uni_l4lb+0xb80 (bpf_prog_2b956549c660136a_uni_l4lb)
        ffffffffc087cbc9 mlx5e_xdp_handle+0xa9 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc0878d87 mlx5e_skb_from_cqe_linear+0xc7 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087a254 mlx5e_handle_rx_cqe+0x64 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087bb48 mlx5e_poll_rx_cq+0x808 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087beee mlx5e_napi_poll+0xde (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffb8f03eaa net_rx_action+0x13a (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb94000e0 __softirqentry_text_start+0xe0 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb886c8c0 irq_exit+0xa0 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb9201908 do_IRQ+0x58 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb9200a0f common_interrupt+0xf (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8ebc022 cpuidle_enter_state+0xb2 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8ebc3c9 cpuidle_enter+0x29 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88960f8 do_idle+0x1b8 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88962c9 cpu_startup_entry+0x19 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8841383 start_secondary+0x143 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88000d4 [unknown] (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
```

### CODE REVIEW

File: src/collapse/perf.rs [hunk 1]
```diff
@@ -514,7 +514,9 @@ impl Folder {
                 if !self.cache_line.is_empty() {
                     func.push_str("_[i]"); // inlined
                 } else if self.opt.annotate_kernel
-                    && (module.starts_with('[') || module.ends_with("vmlinux"))
+                    && (module.starts_with('[')
+                        || module.ends_with(".ko")
+                        || module.contains("vmlinux"))
```
Comment (jonhoo) [inline src/collapse/perf.rs:519]: The updated regex in https://github.com/brendangregg/FlameGraph/pull/232 now allows `vmlinux` anywhere in the string (`vmlinux.*`), so if we want to match the behavior exactly, the `vmlinux` clause would need to be `.contains`.

**Comment by jonhoo** [comment]
It'd be great to include the examples from the upstream PR as tests!
Also, you may want to rebase on the `master` I just pushed :sweat_smile:

**Comment by AnderEnder** [comment]
@jonhoo, can we add regex lib? The draft implementation uses `contains`, while the proper one should check something like `vmlinux([\d\w\-]+)$`

File: src/collapse/perf.rs (hunk 1 above)
Comment (AnderEnder) [reply src/collapse/perf.rs:519]: It is

Comment (jonhoo) [reply src/collapse/perf.rs:519]: You're right. I'm clearly blind.

**Comment by jonhoo** [comment]
I'd still prefer not to bring in regex, since I think these are pretty easy parsing cases:
```rust
fn is_vmlinux(s: &str) -> bool {
  if let Some(vm) = s.rfind("vmlinux") {
    s[vm..].chars().all(|c| c.is_ascii_alphanumeric() || c == '-')
  } else {
    false
  }
}
```

**Comment by AnderEnder** [comment]
@jonhoo,
What is to reason to avoid regex? Rust regex is very efficient.

I made a benchmark of your string function example with regex to test stack example:
Code: [regex_bench.rs](https://gist.github.com/AnderEnder/c9ee0788ce6d0f76cba4f8284e72c6d7)
Benchmark results: [gist](https://gist.github.com/AnderEnder/a3cd84249031583d5040e8dc707d7cd0)

Regex is in about 3 times faster for this particular case.

**Comment by AnderEnder** [comment]
@jonhoo, I put the functions into separate module `matcher`. Can you confirm this is good?

**Comment by jonhoo** [comment]
Yes, that looks great, thanks!

**Comment by jonhoo** [comment]
I'm surprised that this doesn't affect any of the tests.. We don't have any with versioned kernel symbols I suppose? Maybe it'd be good to add a short one? Not sure if you feel like it might be worth it?

**Comment by AnderEnder** [comment]
@jonhoo, It would be good to add this case to test. Unfortunately I was not able to get the example and I do not want to create an artificial one instead.

**Comment by jonhoo** [comment]
@AnderEnder Can't we use the ones in the original issue? Namely
```
        ffffffffb8f6abf0 nf_hook_slow+0x40 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8f74051 ip_local_deliver+0xd1 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8f7412c ip_rcv+0xbc (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
```
and
```
swapper     0 [012] 2281339.604067:   20834821 cycles:
        ffffffffc09ace93 bpf_prog_2b956549c660136a_uni_l4lb+0xb80 (bpf_prog_2b956549c660136a_uni_l4lb)
        ffffffffc087cbc9 mlx5e_xdp_handle+0xa9 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc0878d87 mlx5e_skb_from_cqe_linear+0xc7 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087a254 mlx5e_handle_rx_cqe+0x64 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087bb48 mlx5e_poll_rx_cq+0x808 (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffc087beee mlx5e_napi_poll+0xde (/lib/modules/5.4.14-cloudflare-2020.1.11/kernel/drivers/net/ethernet/mellanox/mlx5/core/mlx5_core.ko)
        ffffffffb8f03eaa net_rx_action+0x13a (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb94000e0 __softirqentry_text_start+0xe0 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb886c8c0 irq_exit+0xa0 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb9201908 do_IRQ+0x58 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb9200a0f common_interrupt+0xf (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8ebc022 cpuidle_enter_state+0xb2 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8ebc3c9 cpuidle_enter+0x29 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88960f8 do_idle+0x1b8 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88962c9 cpu_startup_entry+0x19 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb8841383 start_secondary+0x143 (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
        ffffffffb88000d4 [unknown] (/usr/lib/debug/boot/vmlinux-5.4.14-cloudflare-2020.1.11)
```

**Comment by jonhoo** [comment]
Released in 0.10.1.

**Pull request merged**